src/main/java/io/nats/client/impl/NatsConnection.java

```diff
@@ -712,19 +712,13 @@ void handleCommunicationIssue(Exception io) {
         // Spawn a thread so we don't have timing issues with
         // waiting on read/write threads
         executor.submit(() -> {
-            if (!tryingToConnect.get()) {
-                try {
-                    tryingToConnect.set(true);
-
-                    // any issue that brings us here is pretty serious
-                    // so we are comfortable forcing the close
-                    this.closeSocket(true, true);
-                } catch (InterruptedException e) {
-                    processException(e);
-                    Thread.currentThread().interrupt();
-                } finally {
-                    tryingToConnect.set(false);
-                }
+            try {

```

Comment:
Removing the `tryingToConnect` guard here puts us back in the situation where two threads can be actively closing and reconnecting us at the same time.

The reason why the tests failed is that this `closeSocket` call would put us in `tryingToConnect`.
Then we'd try the reconnect in `closeSocket`:
```java
if (isClosing()) { // isClosing() means we are in the close method or were asked to be
    close();
} else if (wasConnected && tryReconnectIfConnected) {
    reconnect();
}
```

But, `reconnect` has a check if we're not `tryingToConnect` already:
```java
   void reconnect() throws InterruptedException {
        if (!tryingToConnect.get()) {
            try {
                tryingToConnect.set(true);
                reconnectImpl();
            }
            finally {
                tryingToConnect.set(false);
            }
        }
    }
```

So, I would propose to instead change that to `reconnectImpl()`, just like was done in another spot:
```java
if (reconnectOnConnect) {
    timeTraceLogger.trace("trying to reconnect on connect");
    reconnectImpl(); // call the impl here otherwise the tryingToConnect guard will block the behavior
}
```

Have implemented that here which fixes the tests: https://github.com/nats-io/nats.java/pull/1206